Docker build performance optimization

angelej opened the pull request:
This is an [optimization](https://docs.docker.com/build/cache/#optimizing-how-you-use-the-build-cache) for the docker build cache by reordering the layers and combining commands.
With these changes, the docker image has 5 fewer layers and a faster build time.
I hope you like it!

angelej committed:
Docker build performance optimization

ondrejmirtes commented:
Hello, can you please share measurements how faster it is? Also the contents are now a bit different, for example you're no longer doing `rm -rf /var/cache/apk/* /var/tmp/* /tmp/*`.

angelej commented:
Sure! I've replaced `rm -rf /var/cache/apk/* /var/tmp/* /tmp/*` with the `--no-cache` option in the `apk add --no-cache git` statement. The directories are still empty after building the images.

Measuring the performance was a little bit tricky. The docker cli only shows one decimal number and most stages already take 0.0s when loaded from the cache. However, we can assume a faster build time because it has less layers to load from the cache, even if its only a fraction of a second.
But, when the cache gets invalidated e.g. through a new composer version, then the build time is in fact faster.
This happens because once the cache is invalidated, all subsequent Dockerfile commands generate new images and the cache isn't used. Due to this behavior, I moved a few commands above other commands that could invalidate the cache.

I had to hardcode a random composer version for this test in order to simulate a "new" composer version and invalidate the cache at this point in the Dockerfile.

**Old Dockerfile.php.8.0**
![image](https://github.com/phpstan/phpstan/assets/131715596/7246676f-6635-48e6-a944-ef3b89764d4e)

**New Dockerfile.php.8.0**
![image](https://github.com/phpstan/phpstan/assets/131715596/16ec3201-6f23-4823-8de7-9ad521b24fbb)

ondrejmirtes commented:
Alright, thank you, you clearly know what you're doing :)

angelej commented:
> Alright, thank you, you clearly know what you're doing :)

Thanks :)
